grc: Include GRC tests in CTest testing

velichkov opened the pull request:
- Mark failing tests with xfail
- Fix test_extend

Closes: #2680

velichkov commented:
```
        Start 152: grc_tests
152/366 Test #152: grc_tests ..........................................***Failed    0.04 sec
/usr/bin/python3: No module named pytest
```

Would it be possible to install pytest in the buildbots containers? Should I open a PR in gnuradio-buildbot project?

I'm going to add a check for `pytest` in the CMakeList.txt and execute the tests only when present in a subsequent commit.

haakov commented:
related PR: https://github.com/gnuradio/gnuradio/pull/2679

marcusmueller commented:
hm, this silently adds pytest as new (optional) dependency, without documenting it, right?

marcusmueller commented:
Aside from that: this change set exceeds the pretty strict "minimum significant change to require copyright assignment" as of now. Can you please get in contact with @bhilburn and apply for a CLA?

velichkov commented:
Hi @marcusmueller,

> hm, this silently adds pytest as new (optional) dependency, without documenting it, right?

No, pytest is already used in two places. Let me know if I need to update some documentation as well.

https://github.com/gnuradio/gnuradio/blob/267d669eb21c514c18a6ee979f5cf247d251f1ad/grc/tests/test_block_templates.py#L1
https://github.com/gnuradio/gnuradio/blob/267d669eb21c514c18a6ee979f5cf247d251f1ad/grc/tests/test_expr_utils.py#L3

I've added a cmake check for pytest and marked three tests as expected failures. Also I just rebased on top of current master to resolve one merge conflict.

>  Can you please get in contact with @bhilburn and apply for a CLA?

I've signed and sent the CLA and got back a confirmation.

mbr0wn commented on grc/tests/test_expr_utils.py:
```diff
@@ -8,6 +8,7 @@
 expr_getter = operator.itemgetter(1)
 
 
+@pytest.mark.xfail(reason="core/utils/expr_utils.py:97: TypeError: '<' not supported between instances of 'NoneType' and 'str'")
```
Very cool

mbr0wn commented on grc/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+
+if(ENABLE_TESTING)
+    execute_process(
+        COMMAND ${QA_PYTHON_EXECUTABLE} -m pytest --version
```
I think we have a module for this.

velichkov replied:
You mean a cmake module that executes pytest? I'm not able to find such module but I'm going to redo this if you tell me where it is.

```
$ git describe
v3.8.0.0-196-g65a8c6c526
$ grep pytest . -R
./grc/tests/test_block_templates.py:import pytest
./grc/tests/test_block_templates.py:    with pytest.raises(TemplateError):
./grc/tests/test_block_templates.py:    with pytest.raises(TemplateError):
./grc/tests/test_expr_utils.py:import pytest
./grc/tests/test_expr_utils.py:    with pytest.raises(RuntimeError):
```

velichkov replied:
I found the `GR_PYTHON_CHECK_MODULE` and `GR_PYTHON_CHECK_MODULE_RAW`. I'm going to rework this

michaelld replied:
`GR_ADD_TEST`?

michaelld replied:
That's what I can find for what uses QA_PYTHON_EXECUTABLE in any capacity in the current GR codebase.

mbr0wn replied:
GR_PYTHON_CHECK_MODULE is what I meant, sorry, that was a bit cryptic.

velichkov replied:
Done. I've used `GR_PYTHON_CHECK_MODULE_RAW` like in `grc/CMakeLists.txt`

michaelld commented on grc/tests/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+# Copyright 2011-212 Free Software Foundation, Inc.
```
Please fix the dates here to just be 2020

velichkov replied:
Done

michaelld commented on grc/tests/test_generator.py:
```diff
@@ -17,12 +17,15 @@
 # the Free Software Foundation, Inc., 51 Franklin Street,
 # Boston, MA 02110-1301, USA.
 
+import pytest
```
Does `grc/CMakeLists.txt` already check for `pytest`? If not, please add that check with respect to the `ENABLE_TESTS` just in GRC.

velichkov replied:
Yes, but the check is in `grc/tests/CMakeLists.txt`

michaelld replied:
:+1:

velichkov commented:
Hi @michaelld, @mbr0wn,

Just fixed 3 expected failures by setting `GRC_BLOCKS_PATH` environment variable to local blocks directory.

Also note that `pytest` is not installed in the buildbot's containers and the grc's tests are not executed.

michaelld commented:
> Just fixed 3 expected failures by setting `GRC_BLOCKS_PATH` environment variable to local blocks directory.

Nice! Such a simple fix, too!
 
> Also note that `pytest` is not installed in the buildbot's containers and the grc's tests are not executed.

Hmmm ... yeah we need to add `pytest` to the BB containers. Guess we'll need to do testing locally for the moment.

mbr0wn commented:
Yup we know about the missing pytest. I still need confirmation from @marcusmueller that we can add the optional pytest dependency, then we can take care of pytest.

That said, adding pytest to the Dockers won't hurt, so if you like you can submit a PR against gnuradio-buildbot to add pytest.

mbr0wn commented:
@marcusmueller gave a green light. This is OK to merge now.
